test/src/com/xilinx/rapidwright/eco/TestECOTools.java

```diff
@@ -0,0 +1,438 @@
+            }
+        }
+
+        // Check source SPI is swapped
+        Assertions.assertTrue(sourceSitePinInsts.get(0).contains(physNets.get(1).getSource().getSitePinName()));
+        Assertions.assertTrue(sourceSitePinInsts.get(1).contains(physNets.get(0).getSource().getSitePinName()));
+
+        if (FileTools.isVivadoOnPath()) {
+            ReportRouteStatusResult rrs = VivadoTools.reportRouteStatus(design);
+            Assertions.assertEquals(0, rrs.netsWithRoutingErrors);
+            Assertions.assertEquals(2, rrs.unroutedNets);
+        }
+    }
+
+    @Test
+    @Disabled("Currently, ECOTools.removeCell() does not work for hierarchical cells. Specifically, for this testcase " +
+            "exclusively intra-site routes (e.g. 'processor/data_path_loop[4].small_spm.small_spm_ram.spm_ram/DOA') " +
+            "are not removed and appear in the DCP causing Vivado to emit 'placement information for XX sites failed to" +
+            "restore' warnings and cells (e.g. 'your_program/ram_4096x8') to be unplaced.")
+    public void testRemoveCell() {
+        Design design = RapidWrightDCP.loadDCP("picoblaze_ooc_X10Y235.dcp");
+        EDIFNetlist netlist = design.getNetlist();
+        Map<Net, Set<SitePinInst>> deferredRemovals = new HashMap<>();
+
+        EDIFHierCellInst ehciToDelete = netlist.getHierCellInstFromName("processor");
+        List<EDIFHierCellInst> leavesToDelete = netlist.getAllLeafDescendants(ehciToDelete);
+
+        ECOTools.removeCell(design, Collections.singletonList(ehciToDelete), deferredRemovals);
+
+        for (EDIFHierCellInst ehci : leavesToDelete) {
+            String instName = ehci.getFullHierarchicalInstName();
+            // Logical leaf cell not present
+            Assertions.assertNull(netlist.getHierCellInstFromName(instName));
+            // Physical cell not present
+            Assertions.assertNull(design.getCell(instName));
+        }
+
+        // Logical hierarchical cell not present
+        Assertions.assertNull(netlist.getHierCellInstFromName(ehciToDelete.getFullHierarchicalInstName()));
+
+        DesignTools.batchRemoveSitePins(deferredRemovals, true);
+
+        design.writeCheckpoint("/group/zircon2/eddieh/pb.dcp");
+
+        if (FileTools.isVivadoOnPath()) {
+            ReportRouteStatusResult rrs = VivadoTools.reportRouteStatus(design);
+            Assertions.assertEquals(0 /* TODO */, rrs.netsWithRoutingErrors);
+        }
+    }
+
+    @Test
+    public void testRemoveCellLeaf() {
+        Design design = RapidWrightDCP.loadDCP("picoblaze_ooc_X10Y235.dcp");
+        EDIFNetlist netlist = design.getNetlist();
+        Map<Net, Set<SitePinInst>> deferredRemovals = new HashMap<>();
+
+        EDIFHierCellInst ehciToDelete = netlist.getHierCellInstFromName("your_program/ram_4096x8");
+        Assertions.assertTrue(ehciToDelete.getCellType().isLeafCellOrBlackBox());
+
+        ECOTools.removeCell(design, Collections.singletonList(ehciToDelete), deferredRemovals);
```

Comment:
It seems `ECOTools.removeCell()` returns an object.  Should we capture the return value and also test it?

Reply:
`ECOTools.removeCell()` does not return anything.